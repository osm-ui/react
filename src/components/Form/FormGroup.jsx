import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import classnames from 'classnames';

const contexts = ['', 'info', 'success', 'warning', 'error'];

const colorsStyle = props =>
  contexts.reduce((reducedStyles, context) => {
    if (!context) {
      return reducedStyles;
    }

    const labelColors = props.theme.form.label[context];
    const inputColors = props.theme.form.input[context];
    const hintColors = props.theme.form.hint[context];

    return `
      ${reducedStyles}

      &.has-${context} {
        .control-label {
            color: ${labelColors.color};
        }

        .Select-control,
        .form-control {
            background-color: ${inputColors.backgroundColor};
            border-color: ${inputColors.borderColor};
        }

        .help-block {
            background-color: ${hintColors.backgroundColor};
            color: ${hintColors.color};
        }
      }
    `;
  }, '');

const StyledDiv = styled.div`
  ${props => colorsStyle(props)} margin-bottom: 4rem;
`;

const FormGroup = ({ context, className, ...props }) => {
  const classes = classnames(className, {
    'form-group': true,
    [`has-${context}`]: context !== ''
  });

  return <StyledDiv className={classes} {...props} />;
};

FormGroup.contexts = contexts;

FormGroup.propTypes = {
  context: PropTypes.oneOf(contexts),
  className: PropTypes.string
};

FormGroup.defaultProps = {
  context: '',
  className: ''
};

FormGroup.displayName = 'Form.Group';

export default FormGroup;
