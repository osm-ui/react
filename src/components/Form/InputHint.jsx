import styled from 'styled-components';

const InputHint = styled.div.attrs({
  className: 'hint'
})`
  font-size: ${props => props.theme.form.hint.fontSize};
  color: ${props => props.theme.form.hint.color};
  background-color: ${props => props.theme.form.hint.backgroundColor};
  margin: ${props => props.theme.form.hint.margin};
  padding: ${props => props.theme.form.hint.padding};
  transition: all ease-in-out 0.15s;
  opacity: ${props =>
    props.disabled ? props.theme.form.hint.disabledOpacity : 1};
`;

InputHint.displayName = 'Form.InputHint';

export default InputHint;
