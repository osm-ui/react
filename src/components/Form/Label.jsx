import styled from 'styled-components';

const Label = styled.label`
  color: ${p => p.theme.form.label.color};
  background-color: ${p => p.theme.form.label.backgroundColor};
  font-size: ${p => p.theme.form.label.fontSize};
  font-weight: ${p => p.theme.form.label.fontWeight};
  line-height: ${p => p.theme.form.label.lineHeight};
`;

Label.displayName = 'Form.Label';

export default Label;
