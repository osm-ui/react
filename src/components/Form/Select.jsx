import React from 'react';
import styled from 'styled-components';
import 'react-select/dist/react-select.css';
import 'react-virtualized/styles.css';
import 'react-virtualized-select/styles.css';
import VirtualizedSelect from 'react-virtualized-select';

const StyledSelect = styled(VirtualizedSelect)`
  .Select-control {
    opacity: ${props => props.theme.form.input.opacity};
    background-color: ${props => props.theme.form.input.backgroundColor};
    border-color: ${props => props.theme.form.input.borderColor};
    border-width: ${props => props.theme.form.input.borderWidth};
    border-radius: ${props => props.theme.form.input.borderRadius};
    box-shadow: none;
  }

  &.is-focused .Select-control,
  &.is-focused:not(.is-open) > .Select-control {
    opacity: ${props => props.theme.form.input.focusOpacity};
    background-color: ${props => props.theme.form.input.focusBackgroundColor};
    border-color: ${props => props.theme.form.input.focusBorderColor};
    box-shadow: none;
  }

  .Select-placeholder {
    padding-left: 12px;
    line-height: 38px;
  }

  &.Select--single {
    .Select-value {
      padding-left: 12px;
    }

    .Select-value .Select-value-label {
      line-height: 38px;
    }
  }

  .Select-input > input {
    padding: 1px;
    line-height: 32px;
  }

  .Select-menu-outer {
    margin-top: 0;
    border-top-width: 0;
  }

  .VirtualizedSelectOption {
    cursor: default;
    padding: 5px 10px;
  }

  .VirtualizedSelectFocusedOption {
    color: ${props => props.theme.form.select.optionColor};
    background-color: ${props => props.theme.form.select.optionBackgroundColor};
  }
`;

class Select extends React.Component {
  constructor(props) {
    super(props);

    this.state = {};
  }

  render() {
    return (
      <div>
        <StyledSelect
          onChange={selectValue => this.setState({ selectValue })}
          value={this.state.selectValue}
          {...this.props}
        />
      </div>
    );
  }
}

Select.propTypes = {
  ...VirtualizedSelect.PropTypes
};

Select.defaultProps = {
  ...VirtualizedSelect.defaultProps
};

Select.displayName = 'Form.Select';

export default Select;
