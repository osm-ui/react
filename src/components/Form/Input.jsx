import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import classnames from 'classnames';
import InputHint from './InputHint';

const StyledInput = styled.input`
  opacity: ${p => p.theme.form.input.opacity};
  background-color: ${p => p.theme.form.input.backgroundColor};
  border-style: solid;
  border-color: ${p => p.theme.form.input.borderColor};
  border-width: ${p => p.theme.form.input.borderWidth};
  border-radius: ${p => p.theme.form.input.borderRadius};
  box-shadow: ${p => p.theme.form.input.boxShadow};
  padding: 0.6rem 0.8rem;
  width: 100%;
  font-size: ${p => p.theme.form.input.fontSize};
  color: ${p => p.theme.color};

  /* prettier-ignore */
  ${p => (p.hasHint ? 'border-bottom-left-radius: 0;' : '')}
  ${p => (p.hasHint ? 'border-bottom-right-radius: 0;' : '')}

  &:focus {
    opacity: ${p => p.theme.form.input.focusOpacity};
    background-color: ${p => p.theme.form.input.focusBackgroundColor};
    border-color: ${p => p.theme.form.input.focusBorderColor};
    box-shadow: ${p => p.theme.form.input.focusBoxShadow};
  }

  &:disabled {
    opacity: ${p => p.theme.form.input.disabledOpacity};
  }
`;

const Input = ({
  type,
  disabled,
  hint,
  className,
  innerRef,
  style,
  ...props
}) => (
  <div className="input-container" style={style}>
    <StyledInput
      className={classnames(className, 'input')}
      type={type}
      disabled={disabled}
      innerRef={innerRef}
      hasHint={!!hint}
      {...props}
    />
    {!!hint && <InputHint disabled={disabled}>{hint}</InputHint>}
  </div>
);

Input.propTypes = {
  type: PropTypes.string,
  disabled: PropTypes.bool,
  hint: PropTypes.string,
  innerRef: PropTypes.func,
  className: PropTypes.string,
  style: PropTypes.object
};

Input.defaultProps = {
  type: 'text',
  disabled: false,
  hint: '',
  innerRef: null,
  className: '',
  style: null
};

Input.displayName = 'Form.Input';
Input.style = StyledInput;

export default Input;
