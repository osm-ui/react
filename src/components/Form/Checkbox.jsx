import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import Label from './Label';

const Container = styled.div`
  position: relative;
  color: ${props => props.theme.form.label.color};
  background-color: ${props => props.theme.form.label.backgroundColor};
  border-color: ${props => props.theme.form.label.borderColor};
`;

const StyledCheckbox = styled.input.attrs({
  type: 'checkbox',
  checked: props => props.checked,
  disabled: props => props.disabled
})`
  opacity: 0;
  width: 0;
  height: 0;
  margin: 0;

  & + label {
    font-size: ${props => props.theme.form.label.fontSize};
    line-height: ${props => props.theme.form.label.fontSize};
    font-weight: ${props => props.theme.form.label.fontWeight};
    padding-left: 2.5rem;
    cursor: pointer;

    &:before,
    &:after {
      display: inline-block;
      position: absolute;
      left: 0;
      top: 0.4rem;
      border: 1px solid #ccc;
      border-radius: 3px;
      border-color: ${props => props.theme.form.checkbox.borderColor};
      content: '';
      width: 1.7rem;
      height: 1.7rem;
      transition: border 0.15s ease-in-out, color 0.15s ease-in-out;
    }
  }

  &:checked + label:after {
    padding-left: 0.5px;
    content: '\f00c';
    font-family: FontAwesome;
    font-size: 1.4rem;
  }

  &:disabled + label,
  &:disabled + label:after {
    opacity: 0.3;
    cursor: not-allowed;
  }
`;

const Checkbox = ({ id, label, className, ...props }) => {
  return (
    <Container className={className}>
      <StyledCheckbox id={id} {...props} />
      <Label htmlFor={id}>{label}</Label>
    </Container>
  );
};

Checkbox.propTypes = {
  id: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  label: PropTypes.string,
  disabled: PropTypes.bool,
  className: PropTypes.string
};

Checkbox.defaultProps = {
  label: '',
  disabled: false,
  className: ''
};

Checkbox.displayName = 'Form.Checkbox';

export default Checkbox;
