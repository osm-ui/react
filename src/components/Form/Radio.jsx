import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import Label from './Label';

const Container = styled.div`
  position: relative;
  color: ${props => props.theme.form.label.color};
  background-color: ${props => props.theme.form.label.backgroundColor};
  border-color: ${props => props.theme.form.label.borderColor};
  height: ${props => props.theme.form.checkbox[props.size].height};
`;

const StyledRadio = styled.input.attrs({
  type: 'radio',
  checked: props => props.checked,
  disabled: props => props.disabled
})`
  opacity: 0;
  width: 0;
  height: 0;
  margin: 0;

  & + label {
    font-size: ${props => props.theme.form.radio[props.size].size};
    line-height: ${props => props.theme.form.radio[props.size].size};
    font-weight: ${props => props.theme.form.label.fontWeight};
    padding-left: ${props => props.theme.form.radio[props.size].paddingLeft};
    cursor: pointer;
    transition: opacity 0.2s ease-in-out;

    &:before,
    &:after {
      display: inline-block;
      position: absolute;
      left: 0;
      top: ${props => props.theme.form.radio[props.size].top};
      border: 1px solid #ccc;
      border-radius: 50%;
      border-color: ${props => props.theme.form.radio.borderColor};
      content: '';
      width: ${props => props.theme.form.radio[props.size].size};
      height: ${props => props.theme.form.radio[props.size].size};
    }

    &:after {
      border: none;
      padding-left: 0.5px;
      content: ' ';
      width: ${props => props.theme.form.radio[props.size].innerSize};
      height: ${props => props.theme.form.radio[props.size].innerSize};
      top: ${props => props.theme.form.radio[props.size].innerTop};
      left: ${props => props.theme.form.radio[props.size].innerLeft};
      background-color: ${props => props.theme.form.radio.backgroundColor};
      transform: scale(0);
      transition: transform 0.3s cubic-bezier(0.8, -0.33, 0.2, 1.33);
    }
  }

  &:checked + label:after {
    transform: scale(1);
    transition: transform 0.3s cubic-bezier(0.8, -0.33, 0.2, 1.33);
  }

  &:disabled + label,
  &:disabled + label:after {
    opacity: 0.3;
    cursor: not-allowed;
  }
`;

const Radio = ({ id, label, className, ...props }) => {
  return (
    <Container className={className} {...props}>
      <StyledRadio id={id} {...props} />
      <Label htmlFor={id}>{label}</Label>
    </Container>
  );
};

Radio.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  size: PropTypes.oneOf(['sm', 'md', 'lg']),
  label: PropTypes.string,
  disabled: PropTypes.bool,
  className: PropTypes.string
};

Radio.defaultProps = {
  label: '',
  size: 'md',
  disabled: false,
  className: ''
};

Radio.displayName = 'Form.Radio';

export default Radio;
