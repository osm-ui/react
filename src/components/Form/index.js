import Form from './Form';
import FormGroup from './FormGroup';
import Label from './Label';
import Input from './Input';
import Textarea from './Textarea';
import Select from './Select';
import Checkbox from './Checkbox';
import Radio from './Radio';

Form.Group = FormGroup;
Form.Label = Label;
Form.Input = Input;
Form.Textarea = Textarea;
Form.Select = Select;
Form.Checkbox = Checkbox;
Form.Radio = Radio;

export default Form;
