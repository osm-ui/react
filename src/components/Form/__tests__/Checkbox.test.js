import 'jest-styled-components';
import { snapshotWithElementChildren } from 'helpers/tests';
import Form from '../index';

describe('When using snapshots', () => {
  it('Checkbox should render with an element children', () =>
    snapshotWithElementChildren(Form.Checkbox, { id: '1', value: 'myValue' }));
});
