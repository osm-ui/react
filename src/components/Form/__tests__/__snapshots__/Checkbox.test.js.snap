// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When using snapshots Checkbox should render with an element children 1`] = `
<Default__StyledSpan>
  <Form.Checkbox
    className=""
    disabled={false}
    id="1"
    key=".0"
    label=""
    size="md"
    value="myValue"
  >
    <div />
  </Form.Checkbox>
</Default__StyledSpan>
`;
